control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs

```diff
@@ -130,61 +136,139 @@ async fn nexus_reconciler(
+                let _ = handle_child_rebuild(nexus, child, &nexus_state, context).await;
             }
         }
-        .instrument(span)
+    }
+    Ok(PollerState::Idle)
+}
+
+/// Checks rebuild policy for the child. Depending on the policy performs the either waits for
+/// child to comeback for specified duration or does Full rebuild straightaway.
+async fn handle_child_rebuild(
+    nexus_spec: &mut OperationGuardArc<NexusSpec>,
+    child: &Child,
+    nexus: &Nexus,
+    context: &PollContext,
+) -> Result<(), SvcError> {
+    let wait_duration = RuleSet::faulted_child_wait(nexus, context.registry());
+    if wait_duration.is_zero() || child.state_reason == ChildStateReason::RebuildFailed {
+        info!(%child.uri, "Start full rebuild for child");
+        faulted_children_remover(nexus_spec, &child.uri, context).await?
+    } else if is_time_elapsed(child.faulted_at, wait_duration, &child.uri).await {
+        info!(%child.uri, "Start full rebuild for child,  Wait time has elapsed");
+        faulted_children_remover(nexus_spec, &child.uri, context).await?;
+    } else if get_child_replica(nexus_spec.as_ref(), child, context)
         .await
+        .is_err()
+    {
+        info!(%child.uri,"Replica is still not online for child");
+    } else {
+        info!(%child.uri, "Replica for child is Online, starting partial rebuild");
+        if let Err(error) = initialize_partial_rebuild(&nexus.uuid, &child.uri, context).await {
+            info!(%child.uri, "Initializing Full rebuild as partial rebuild failed for child, failed with : {:?}", error);
+            faulted_children_remover(nexus_spec, &child.uri, context).await?
+        }
     }
+    Ok(())
+}
 
-    PollResult::Ok(PollerState::Idle)
+/// Removes child from nexus children list to initiate Full rebuild of child.
+#[tracing::instrument(skip(nexus, context), level = "trace", fields(nexus.uuid = %nexus.uuid(), request.reconcile = true))]
+pub(super) async fn faulted_children_remover(
+    nexus: &mut OperationGuardArc<NexusSpec>,
+    child: &ChildUri,
+    context: &PollContext,
+) -> Result<(), SvcError> {
+    let nexus_uuid = nexus.uuid();
+    let nexus_state = context.registry().nexus(nexus_uuid).await?;
+
+    info!(%child, "Attempting to remove faulted child");
+    nexus
+        .remove_child_by_uri(context.registry(), &nexus_state, child, true)
+        .await?;
+
+    info!(%child ,"Successfully removed faulted child");
+    Ok(())
+}
+
+/// Returns true if time elapsed from child fault time is greater or equal to wait time duration.
+async fn is_time_elapsed(
+    fault_time: Option<SystemTime>,
+    wait_duration: Duration,
+    uri: &ChildUri,
+) -> bool {
+    if let Some(fault_time) = fault_time {
+        if let Ok(elapsed) = fault_time.elapsed() {
+            info!(
+                child.uri = %uri,
+                "waited {:?} for child to comeback. Wait time is {:?}",
+                elapsed, wait_duration
+            );
```

Comment:
don't need this since we have logs above

Reply:
Here it prints how much we have waited for the child during every attempt.

Reply:
I don't think that's necessary.